extensions/vscode/src/views/lessonsTree.ts

```diff
@@ -1,114 +1,98 @@
-
-export function getLessonsTreeDataProvider() {
-  return lessonsTreeDataProvider;
-}
+export class LessonsTreeDataProvider implements vscode.TreeDataProvider<Node>, vscode.Disposable {
+  private _tutorial!: Node;
+  private _tutorialName: string;
+  private _onDidChangeTextDocumentDisposable: vscode.Disposable;
+  private _onDidChangeTreeData: vscode.EventEmitter<Node | undefined> = new vscode.EventEmitter<Node | undefined>();
 
-export function setLessonsTreeDataProvider(provider: LessonsTreeDataProvider) {
-  lessonsTreeDataProvider = provider;
-}
-
-export class LessonsTreeDataProvider implements vscode.TreeDataProvider<Lesson> {
-  private _lessons: Lesson[] = [];
-  private _onDidChangeTreeData: vscode.EventEmitter<Lesson | undefined> = new vscode.EventEmitter<Lesson | undefined>();
-  readonly onDidChangeTreeData: vscode.Event<Lesson | undefined> = this._onDidChangeTreeData.event;
+  readonly onDidChangeTreeData: vscode.Event<Node | undefined> = this._onDidChangeTreeData.event;
 
   constructor(
     private readonly _workspaceRoot: vscode.Uri,
     private _context: vscode.ExtensionContext,
   ) {
-    this._loadLessons();
-  }
-
-  private _loadLessons(): void {
-    try {
-      const tutorialFolderPath = vscode.Uri.joinPath(this._workspaceRoot, 'src', 'content', 'tutorial').fsPath;
-      this._lessons = this._loadLessonsFromFolder(tutorialFolderPath);
-    } catch {
-      // do nothing
-    }
-  }
-
-  private _loadLessonsFromFolder(folderPath: string): Lesson[] {
-    const lessons: Lesson[] = [];
-    const files = fs.readdirSync(folderPath);
+    this._tutorialName = path.basename(_workspaceRoot.path);
 
-    for (const file of files) {
-      const filePath = path.join(folderPath, file);
-      const stats = fs.statSync(filePath);
+    let timeoutId: ReturnType<typeof setTimeout>;
+    let loading = false;
 
-      if (stats.isDirectory()) {
-        const lessonName = path.basename(filePath);
-        const subLessons = this._loadLessonsFromFolder(filePath);
-        const lesson = new Lesson(lessonName, filePath, subLessons);
-
-        // check if the folder directly includes one of the metadata files
-        const folderFiles = fs.readdirSync(filePath);
-        const metadataFile = folderFiles.find((folderFile) => metadataFiles.includes(folderFile));
+    this._onDidChangeTextDocumentDisposable = vscode.workspace.onDidChangeTextDocument((documentChange) => {
+      if (loading || !METADATA_FILES.has(path.basename(documentChange.document.fileName))) {
+        return;
+      }
 
-        if (metadataFile) {
-          const metadataFilePath = path.join(filePath, metadataFile);
-          const metadataFileContent = fs.readFileSync(metadataFilePath, 'utf8');
-          const parsedContent = grayMatter(metadataFileContent);
+      clearTimeout(timeoutId);
 
-          lesson.name = parsedContent.data.title;
+      timeoutId = setTimeout(async () => {
+        loading = true;
+        await this.refresh();
```

Comment:
When a file is changed, we throw away everything and recalculate everything (lazily on a needed basis though). 

This is pretty inefficient but very simple.

Ideally, I'd want to do something more fine-grained where we only update the element of the tree that need to. This should improve perf on really big tutorials with many many markdown files.